### PR TITLE
Add .pre-commit hooks

### PR DESCRIPTION
**Pull request by janiversen**
This PR adds but do not activate the pre-commit hooks, similar (but reduced) to HA.

It runs black/flake8/isort/mypy

to activate it run:
pre-commit install

However especially black and flake8 made a lot of corrections, so these needs to be merged first, and then githooks will be activated (saving time by not getting surprises from CI).

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+          - --safe
+          - --quiet
+        files: ^((homeassistant|script|tests)/.+)?[^/]+\.py$
+  - repo: https://gitlab.com/pycqa/flake8
+    rev: 3.9.2
+    hooks:
+      - id: flake8
+        stages: [manual]
+        additional_dependencies:
+          - pycodestyle==2.7.0
```
Comment (MartinHjelmare) [inline .pre-commit-config.yaml:17]: Please make sure that the tox run is using the same linting as pre-commit, by running pre-commit in tox.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -0,0 +1,41 @@
+repos:
+  - repo: https://github.com/psf/black
+    rev: 21.7b0
+    hooks:
+      - id: black
+        stages: [manual]
+        args:
+          - --safe
+          - --quiet
+        files: ^((homeassistant|script|tests)/.+)?[^/]+\.py$
```
[the remark this answers is not in the document]
Comment (janiversen) [reply .pre-commit-config.yaml:10]: Yes of course, sorry I forgot to convert this to draft status. I just wanted to signal 2 things:
A) I am working on getting a proper githook, to save work for me and others who commit
B) Testing the code (Black  make changes to a larger number of files).

Before this gets merged, I need to have to code stable in relation to black etc.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (janiversen) [reply .pre-commit-config.yaml:17]: That is a very good idea, will do.

File: .pre-commit-config.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (janiversen) [reply .pre-commit-config.yaml]: Done, all are still on manual.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (janiversen) [reply .pre-commit-config.yaml:19]: I have to research this a bit more.
- seems flake8 do not run in CI (it comes with a number of “suggestions” when I run it local)
- black runs with —check, where in pre-commit it make changes to the code.

**Comment by janiversen** [comment]
I have now test all entries with:
```
pre-commit run <id> --hook-stage manual --all-files
```
And also
```
pre-commit run --all-files
```
Pytest/black/mypy are active for commit, while flake8/isort/pylint are manual until the sources are updated.

For some reason CI did not like running pre-commit. The different at the moment are:
CI/tox —>  black —check ./
Pre-commit —> black —safe —quiet 
And I would really prefer not to touch tox.ini in this PR.

**Comment by MartinHjelmare** [comment]
It's ok to have black check in CI and write in pre-commit. What's important is that the lint result is exactly the same in both places, ie that we fail linting consistently with exactly the same settings.

Eg, there are plugins used for flake8 in the pre-commit settings. Please check and make sure that they are also applied for tox.

**Comment by janiversen** [comment]
Finally I understood where you were heading, that makes much more sense.

I have add pyproject.toml ("stolen" from HA), and now pylint, black, isort runs identical. flake8 is still missing settings in pyproject.toml (no idea why we do not have them in HA).

I will search for common settings when I start doing the manual changes needed for flake8 (like docstrings).

**Comment by KapJI** [comment]
Flake8 config is in `setup.cfg`: https://github.com/home-assistant/core/blob/dev/setup.cfg
`pyproject.toml` support in flake8 is fairly new but you can start using it there.

**Comment by janiversen** [comment]
thanks a lot, will add it and test it, before somebody approves this PR

**Comment by janiversen** [comment]
Added setup from HA to pyproject.toml, and the number of flagged lines is reduced a lot.

@MartinHjelmare as far as I can see pre-commit and CI find the same problems.

**Comment by janiversen** [comment]
Time to sleep! flake8 was already in setup.cfg (and it does not work yet in pyproject.toml).

**Comment by KapJI** [comment]
I checked again and unfortunately flake8 still doesn't support `pyproject.toml`, it's one of the last tools which doesn't support it yet. So `setup.cfg` is the only option for now.

File: script/run-pre-commit.sh [hunk 3]
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env sh
+set -eu
+
+exec "$@"
```
Comment (KapJI) [inline script/run-pre-commit.sh:4]: What's the point of this script?

`set -e` just exists on error, but here is one one command anyway
`set -u` errors on unset variables but none of them is used here

Comment (janiversen) [reply script/run-pre-commit.sh:4]: The point is that without it pre-commit cannot find mypy
```
  entry: mypy
```
fails with "unable to find mypy" (not copied)
whereas 
```
entry: run-pre-commit.sh mypy
```
Works.
I spent a good hour trying to understand why. The other tools works perfect and mypy works perfect on the command line, just not in pre-commit.

**Comment by janiversen** [comment]
Moved black file settings from .pre-commit.yml to pyproject.toml, to have the same setting when committing and running CI.

File: pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    "HomeAssistantError",
+]
+
+[tool.pylint.TYPING]
+runtime-typing = false
+
+[tool.pylint.CODE_STYLE]
+max-line-length-suggestions = 72
+
+[tool.pytest.ini_options]
```
Comment (MartinHjelmare) [inline pyproject.toml:134]: Is this needed?

File: pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+# any too bad. Override on command line as appropriate.
+jobs = 2
+init-hook='from pylint.config.find_default_config_files import find_default_config_files; from pathlib import Path; import sys; sys.path.append(str(Path(Path(list(find_default_config_files())[0]).parent, "pylint/plugins")))'
+load-plugins = [
+    "pylint.extensions.code_style",
+    "pylint.extensions.typing",
+    "pylint_strict_informational",
+    "hass_constructor",
+    "hass_imports",
+    "hass_logger",
```
Comment (MartinHjelmare) [inline pyproject.toml:35]:
```suggestion
```

File: pyproject.toml [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+]
+persistent = false
+extension-pkg-allow-list = [
+    "av.audio.stream",
+    "av.stream",
+    "ciso8601",
+    "cv2",
+]
```
Comment (MartinHjelmare) [inline pyproject.toml:43]:
```suggestion
```

File: pyproject.toml [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+[tool.pylint.BASIC]
+class-const-naming-style = "any"
+good-names = [
+    "_",
+    "ev",
```
Comment (MartinHjelmare) [inline pyproject.toml:49]:
```suggestion
```

File: pyproject.toml [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+    "ex",
```
Comment (MartinHjelmare) [inline pyproject.toml:50]:
```suggestion
```

File: pyproject.toml [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+# too-many-ancestors - it's too strict.
+# wrong-import-order - isort guards this
+# consider-using-f-string - str.format sometimes more readable
+# ---
+# Enable once current issues are fixed:
+# consider-using-namedtuple-or-dataclass (Pylint CodeStyle extension)
+# consider-using-assignment-expr (Pylint CodeStyle extension)
+disable = [
+    "format",
+    "abstract-class-little-used",
```
Comment (MartinHjelmare) [inline pyproject.toml:81]: Please remove the ones we don't need now.

File: pyproject.toml [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+    "use-symbolic-message-instead",
+]
+
+[tool.pylint.REPORTS]
+score = false
+
+[tool.pylint.TYPECHECK]
+ignored-classes = [
+    "_CountingAttr",  # for attrs
+]
```
Comment (MartinHjelmare) [inline pyproject.toml:116]:
```suggestion
```

File: pyproject.toml [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+
+[tool.pylint.FORMAT]
+expected-line-ending-format = "LF"
+
+[tool.pylint.EXCEPTIONS]
+overgeneral-exceptions = [
+    "BaseException",
+    "Exception",
+    "HomeAssistantError",
```
Comment (MartinHjelmare) [inline pyproject.toml:125]:
```suggestion
```

**Review by MartinHjelmare — COMMENTED** [review]
I don't see any changes on what plugins flake8 runs in tox. We need to either remove those plugins from pre-commit or include them in our `requirements_test.txt`.

File: pyproject.toml (hunk 4 above)
Comment (janiversen) [reply pyproject.toml:116]: Flake8 plugins are moved from .pre-commit-config.yaml to requirements_test.txt

yes, pytest is included .pre-commit-config.yaml and we want it to run identical in tox and pre-commit.

File: pyproject.toml (hunk 5 above)
Comment (janiversen) [reply pyproject.toml]: Done

File: pyproject.toml (hunk 6 above)
Comment (janiversen) [reply pyproject.toml]: Done

File: pyproject.toml (hunk 7 above)
Comment (janiversen) [reply pyproject.toml]: Done but this is a difference to HA.

File: pyproject.toml (hunk 8 above)
Comment (janiversen) [reply pyproject.toml]: Done but this is a difference to HA.

File: pyproject.toml (hunk 10 above)
Comment (janiversen) [reply pyproject.toml]: Done

File: pyproject.toml (hunk 11 above)
Comment (janiversen) [reply pyproject.toml]: done

File: pyproject.toml (hunk 9 above)
Comment (janiversen) [reply pyproject.toml]: I have commented out a few, but that is something I would like to fine-tune when I activate pylint (currently there are pylint errors, which I want to control).

**Comment by janiversen** [comment]
@MartinHjelmare the plugins for flake8 is now loaded in both tox and pre-commit, however that lead to errors in CI, so I had to disable some error codes to have a clean CI.

Changing the code in order to activate the codes again are on the top of my list.

File: pyproject.toml (hunk 7 above)
Comment (MartinHjelmare) [reply pyproject.toml]: We don't really care about what Home Assistant does.

**Comment by MartinHjelmare** [comment]
Please solve the merge conflict.

**Comment by janiversen** [comment]
Done (it was caused because you merged the other PR, thanks for that).

File: pyproject.toml [hunk 12]
```diff
@@ -12,3 +17,107 @@ forced_separate = [
...
+# consider-using-namedtuple-or-dataclass (Pylint CodeStyle extension)
+# consider-using-assignment-expr (Pylint CodeStyle extension)
+disable = [
+    "format",
+#    "abstract-class-little-used",
+    "abstract-method",
+    "cyclic-import",
+    "duplicate-code",
+#    "inconsistent-return-statements",
+#    "locally-disabled",
```
Comment (MartinHjelmare) [inline pyproject.toml:74]: We'll probably need locally-disabled sooner or later, so that we can disable some warnings locally.

**Review by MartinHjelmare — APPROVED** [review]
Looks good!

File: pyproject.toml (hunk 12 above)
Comment (janiversen) [reply pyproject.toml:74]: OK, I will add that to a future PR.

Comment (janiversen) [reply pyproject.toml:74]: Thanks for merging.

**Pull request merged**